File: eventual-consistency-support-core/src/main/java/org/xyattic/eventual/consistency/support/core/utils/SpelUtils.java
package org.xyattic.eventual.consistency.support.core.utils;

import com.google.common.collect.Maps;
import org.springframework.beans.factory.BeanFactory;
import org.springframework.context.expression.BeanExpressionContextAccessor;
import org.springframework.context.expression.BeanFactoryAccessor;
import org.springframework.context.expression.BeanFactoryResolver;
import org.springframework.context.expression.EnvironmentAccessor;
import org.springframework.context.expression.MapAccessor;
import org.springframework.core.convert.ConversionService;
import org.springframework.core.env.Environment;
import org.springframework.expression.*;
import org.springframework.expression.common.TemplateParserContext;
import org.springframework.expression.spel.SpelParserConfiguration;
import org.springframework.expression.spel.standard.SpelExpressionParser;
import org.springframework.expression.spel.support.StandardEvaluationContext;
import org.springframework.expression.spel.support.StandardTypeConverter;
import org.springframework.expression.spel.support.StandardTypeLocator;

import java.util.Map;

/**
 * @author wangxing
 * @create 2018/4/2
 */
public class SpelUtils {

    private static final SpelExpressionParser SPEL_EXPRESSION_PARSER;
    private static final ParserContext PARSER_CONTEXT = new TemplateParserContext();
    private static final BeanResolver BEAN_RESOLVER = new BeanFactoryResolver(SpringBeanUtils.getBeanFactory());
    private static final CustomPropertyAccessor CUSTOM_PROPERTY_ACCESSOR = new CustomPropertyAccessor();
    private static final BeanExpressionContextAccessor BEAN_EXPRESSION_CONTEXT_ACCESSOR = new BeanExpressionContextAccessor();
    private static final BeanFactoryAccessor BEAN_FACTORY_ACCESSOR = new BeanFactoryAccessor();
    private static final MapAccessor MAP_ACCESSOR = new MapAccessor();
    private static final EnvironmentAccessor ENVIRONMENT_ACCESSOR = new EnvironmentAccessor();
    private static final TypeLocator TYPE_LOCATOR = new StandardTypeLocator(SpringBeanUtils.getBeanFactory().getBeanClassLoader());
    private static TypeConverter TYPE_CONVERTER;

    static {
        final SpelParserConfiguration configuration = new SpelParserConfiguration();
        SPEL_EXPRESSION_PARSER = new SpelExpressionParser(configuration);
        final ConversionService conversionService = SpringBeanUtils.getBeanFactory().getConversionService();
        if (conversionService != null) {
            TYPE_CONVERTER = new StandardTypeConverter(conversionService);
        }
    }

    private static String resolve(final String value) {
        return PropertyPlaceholder.resolvePlaceholders(value);
    }

    public static <T> T parse(final String expressionString, final Map<String, Object> rootMap) {
        return parse(expressionString, rootMap, null);
    }

    public static <T> T parse(final String expressionString) {
        return parse(expressionString, null, null);
    }

    public static <T> T parse(final String expressionString, Map<String, Object> rootMap, final Class<T> valClass) {
        rootMap = rootMap == null ? Maps.newHashMap() : rootMap;
        return SPEL_EXPRESSION_PARSER
                .parseExpression(resolve(expressionString), PARSER_CONTEXT)
                .getValue(getEvaluationContext(rootMap), valClass);
    }

    public static <T> T parse(final String expressionString, final Class<T> valClass) {
        return parse(expressionString, null, valClass);
    }

    private static StandardEvaluationContext getEvaluationContext(final Map<String, Object> rootMap) {
        final StandardEvaluationContext evaluationContext = new StandardEvaluationContext();
        final CustomRootObject rootObject = new CustomRootObject();
        rootObject.map = rootMap;
        rootObject.beanFactory = SpringBeanUtils.getBeanFactory();
        rootObject.environment = SpringContextUtils.getEnvironment();
        evaluationContext.setRootObject(rootObject);
        evaluationContext.setBeanResolver(BEAN_RESOLVER);
        evaluationContext.addPropertyAccessor(CUSTOM_PROPERTY_ACCESSOR);
        evaluationContext.addPropertyAccessor(BEAN_EXPRESSION_CONTEXT_ACCESSOR);
        evaluationContext.addPropertyAccessor(BEAN_FACTORY_ACCESSOR);
        evaluationContext.addPropertyAccessor(MAP_ACCESSOR);
        evaluationContext.addPropertyAccessor(ENVIRONMENT_ACCESSOR);
        evaluationContext.setTypeLocator(TYPE_LOCATOR);
        if (TYPE_CONVERTER != null) {
            evaluationContext.setTypeConverter(TYPE_CONVERTER);
        }
        return evaluationContext;
    }

    public static class CustomRootObject {

        private Map<String, Object> map = Maps.newHashMap();
        private BeanFactory beanFactory;
        private Environment environment;

        public Map<String, Object> getMap() {
            return this.map;
        }

        public void setMap(final Map<String, Object> map) {
            this.map = map;
        }

        public BeanFactory getBeanFactory() {
            return this.beanFactory;
        }

        public void setBeanFactory(final BeanFactory beanFactory) {
            this.beanFactory = beanFactory;
        }

        public Environment getEnvironment() {
            return this.environment;
        }

        public void setEnvironment(final Environment environment) {
            this.environment = environment;
        }
    }

    public static class CustomPropertyAccessor implements PropertyAccessor {

        @Override
        public Class<?>[] getSpecificTargetClasses() {
            return new Class[]{CustomRootObject.class};
        }

        @Override
        public boolean canRead(final EvaluationContext context, final Object target, final String name) throws AccessException {
            if (target instanceof CustomRootObject) {
                final CustomRootObject rootObject = (CustomRootObject) target;
                return rootObject.map.containsKey(name) || rootObject.beanFactory.containsBean(name) || rootObject.environment.containsProperty(name);
            }
            return false;
        }

        @Override
        public TypedValue read(final EvaluationContext context, final Object target, final String name) throws AccessException {
            final Object val;
            final CustomRootObject rootObject = (CustomRootObject) target;
            if (rootObject.map.containsKey(name)) {
                val = rootObject.map.get(name);
            } else if (rootObject.beanFactory.containsBean(name)) {
                val = rootObject.beanFactory.getBean(name);
            } else {
                val = rootObject.environment.getProperty(name);
            }
            return new TypedValue(val);
        }

        @Override
        public boolean canWrite(final EvaluationContext context, final Object target, final String name) throws AccessException {
            return false;
        }

        @Override
        public void write(final EvaluationContext context, final Object target, final String name, final Object newValue) throws AccessException {
            throw new AccessException("不支持的操作");
        }
    }

}

File: eventual-consistency-support-core/src/main/java/org/xyattic/eventual/consistency/support/core/utils/SpelUtils.kt
package org.xyattic.eventual.consistency.support.core.utils

import org.springframework.beans.factory.BeanFactory
import org.springframework.context.expression.*
import org.springframework.core.env.Environment
import org.springframework.expression.*
import org.springframework.expression.common.TemplateParserContext
import org.springframework.expression.spel.SpelParserConfiguration
import org.springframework.expression.spel.standard.SpelExpressionParser
import org.springframework.expression.spel.support.StandardEvaluationContext
import org.springframework.expression.spel.support.StandardTypeConverter
import org.springframework.expression.spel.support.StandardTypeLocator

/**
 * @author wangxing
 * @create 2018/4/2
 */
object SpelUtils {
    private var SPEL_EXPRESSION_PARSER: SpelExpressionParser
    private val PARSER_CONTEXT: ParserContext = TemplateParserContext()
    private val BEAN_RESOLVER: BeanResolver = BeanFactoryResolver(SpringBeanUtils.beanFactory)
    private val CUSTOM_PROPERTY_ACCESSOR = CustomPropertyAccessor()
    private val BEAN_EXPRESSION_CONTEXT_ACCESSOR = BeanExpressionContextAccessor()
    private val BEAN_FACTORY_ACCESSOR = BeanFactoryAccessor()
    private val MAP_ACCESSOR = MapAccessor()
    private val ENVIRONMENT_ACCESSOR = EnvironmentAccessor()
    private val TYPE_LOCATOR: TypeLocator = StandardTypeLocator(SpringBeanUtils.beanFactory.beanClassLoader)
    private var TYPE_CONVERTER: TypeConverter? = null
    private fun resolve(value: String): String {
        return PropertyPlaceholder.resolvePlaceholders(value)
    }

    fun <T> parse(expressionString: String, rootMap: Map<String?, Any?>?): T? {
        return parse(expressionString, rootMap, null)
    }

    fun <T> parse(expressionString: String): T? {
        return parse(expressionString, null, null)
    }

    fun <T> parse(expressionString: String, rootMap: Map<String?, Any?>?, valClass: Class<T>?): T? {
        var map = rootMap
        map = map ?: hashMapOf()
        return SPEL_EXPRESSION_PARSER
                .parseExpression(resolve(expressionString), PARSER_CONTEXT)
                .getValue(getEvaluationContext(map), valClass)
    }

    fun <T> parse(expressionString: String, valClass: Class<T>?): T? {
        return parse(expressionString, null, valClass)
    }

    private fun getEvaluationContext(rootMap: Map<String?, Any?>?): StandardEvaluationContext {
        val evaluationContext = StandardEvaluationContext()
        val rootObject = CustomRootObject()
        rootObject.map = rootMap ?: hashMapOf()
        rootObject.beanFactory = SpringBeanUtils.beanFactory
        rootObject.environment = SpringContextUtils.environment
        evaluationContext.setRootObject(rootObject)
        evaluationContext.beanResolver = BEAN_RESOLVER
        evaluationContext.addPropertyAccessor(CUSTOM_PROPERTY_ACCESSOR)
        evaluationContext.addPropertyAccessor(BEAN_EXPRESSION_CONTEXT_ACCESSOR)
        evaluationContext.addPropertyAccessor(BEAN_FACTORY_ACCESSOR)
        evaluationContext.addPropertyAccessor(MAP_ACCESSOR)
        evaluationContext.addPropertyAccessor(ENVIRONMENT_ACCESSOR)
        evaluationContext.typeLocator = TYPE_LOCATOR
        if (TYPE_CONVERTER != null) {
            evaluationContext.typeConverter = TYPE_CONVERTER!!
        }
        return evaluationContext
    }

    class CustomRootObject {

        var map: Map<String?, Any?> = hashMapOf()

        lateinit var beanFactory: BeanFactory

        lateinit var environment: Environment

    }

    class CustomPropertyAccessor : PropertyAccessor {
        override fun getSpecificTargetClasses(): Array<Class<*>> {
            return arrayOf(CustomRootObject::class.java)
        }

        override fun canRead(context: EvaluationContext, target: Any, name: String): Boolean {
            if (target is CustomRootObject) {
                val rootObject = target
                return rootObject.map.containsKey(name) || rootObject.beanFactory.containsBean(name) || rootObject.environment.containsProperty(name)
            }
            return false
        }

        override fun read(context: EvaluationContext, target: Any, name: String): TypedValue {
            val `val`: Any?
            val rootObject = target as CustomRootObject
            if (rootObject.map.containsKey(name)) {
                `val` = rootObject.map[name]
            } else if (rootObject.beanFactory.containsBean(name)) {
                `val` = rootObject.beanFactory.getBean(name)
            } else {
                `val` = rootObject.environment.getProperty(name)
            }
            return TypedValue(`val`)
        }

        override fun canWrite(context: EvaluationContext, target: Any, name: String): Boolean {
            return false
        }

        override fun write(context: EvaluationContext, target: Any, name: String, newValue: Any) {
            throw AccessException("不支持的操作")
        }
    }

    init {
        val configuration = SpelParserConfiguration()
        SPEL_EXPRESSION_PARSER = SpelExpressionParser(configuration)
        val conversionService = SpringBeanUtils.beanFactory.conversionService
        if (conversionService != null) {
            TYPE_CONVERTER = StandardTypeConverter(conversionService)
        }
    }
}